src/hotspot/os/linux/memMapPrinter_linux.cpp

```diff
@@ -27,58 +27,291 @@
+    st->print_cr("               rss: %zu (" PROPERFMT ")", _rss, PROPERFMTARGS(_rss));
+    st->print_cr("         committed: %zu (" PROPERFMT ")", _committed, PROPERFMTARGS(_committed));
+    st->print_cr("            shared: %zu (" PROPERFMT ")", _shared, PROPERFMTARGS(_shared));
+    st->print_cr("       swapped out: %zu (" PROPERFMT ")", _swapped_out, PROPERFMTARGS(_swapped_out));
+    st->print_cr("         using thp: %zu (" PROPERFMT ")", _thp, PROPERFMTARGS(_thp));
+    st->print_cr("           hugetlb: %zu (" PROPERFMT ")", _hugetlb, PROPERFMTARGS(_hugetlb));
+    st->print_cr("By page size:");
+    _pagesizes.print_on(session);
+  }
 };
 
-void MemMapPrinter::pd_print_header(outputStream* st) {
-  st->print_cr("size          prot offset  What");
+static void print_legend(MappingPrintSession& session) {
+  outputStream* st = session.out();
+  st->print_cr("from, to, size: address range and size");
+  st->print_cr("rss:            resident set size");
+  st->print_cr("pgsz:           page size");
+  st->print_cr("notes:          mapping information");
+  st->print_cr("                    shared: mapping is shared");
+  st->print_cr("                     nores: mapping uncommitted (no swap space reserved)");
+  st->print_cr("                      swap: mapping partly or completely swapped out");
+  st->print_cr("                       thp: mapping uses THP");
+  st->print_cr("                     thpel: mapping is eligible for THP");
+  st->print_cr("                    thpadv: mapping is THP-madvised");
+  st->print_cr("                     nothp: mapping will not THP");
+  st->print_cr("                      huge: mapping uses hugetlb pages");
+  st->print_cr("vm info:        VM information (requires NMT)");
+  session.print_nmt_flag_legend(16);
+  st->print_cr("file:           file mapped, if mapping is not anonymous");
+}
+
+static void print_header(MappingPrintSession& session) {
+  outputStream* st = session.out();
+  //            .         .         .         .         .         .         .         .         .         .         .
+  //            01234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789
+  //            0x0000000414000000 - 0x0000000453000000 1234567890 1234567890 1234567890 16g  rw-p thp,thpadv       JAVAHEAP /shared/tmp.txt
+  st->print_cr("from                 to                       size        rss    hugetlb pgsz prot notes            vm-info file");
 }
 
-void MemMapPrinter::pd_iterate_all_mappings(MappingPrintClosure& closure) {
-  FILE* f = os::fopen("/proc/self/maps", "r");
+void MemMapPrinter::pd_print_all_mappings(MappingPrintSession& session) {
+  outputStream* st = session.out();
+
+  const bool print_individual_mappings = !session.summary_only();
+
+  if (!session.summary_only()) {
+    print_legend(session);
+    st->cr();
+    print_header(session);
+  }
+
+  FILE* f = os::fopen("/proc/self/smaps", "r");
```

Comment:
We have seen that reading the smaps file can be extremely bad for the latency of the process. We've seen multi-seconds hangs because of external tools reading smaps. If we add a tool like this it would be good to add a big warning somewhere.

Reply:
That is a good point, and I was afraid of that myself. 

I think I will revise this coding and the command a "detail" mode, and read smaps (with a warning in the "cost" jcmd category) only in details mode. In normal mode, "maps" is sufficient. And in summary mode I can get probably away with reading stat.